docs/BR.md

```diff
@@ -1756,153 +1754,1150 @@ The CA SHALL enforce multi-factor authentication for all accounts capable of dir
+| `countryName`                  | MUST            | The two-letter ISO 3166-1 country code for the country associated with the Subject. If a Country is not represented by an official ISO 3166-1 country code, the CA MUST specify the ISO 3166-1 user-assigned code of `XX`, indicating that an official ISO 3166-1 alpha-2 code has not been assigned. | [Section 3.2.3](#323-authentication-of-individual-identity) |
+| `stateOrProvinceName`          | MUST / MAY      | MUST be present if `localityName` is absent, MAY be present otherwise. If present, MUST contain the Subject's state or province information. | [Section 3.2.3](#323-authentication-of-individual-identity) |
+| `localityName`                 | MUST / MAY      | MUST be present if `stateOrProvinceName` is absent, MAY be present otherwise. If present, MUST contain the Subject's locality information. | [Section 3.2.3](#323-authentication-of-individual-identity) |
+| `postalCode`                   | NOT RECOMMENDED | If present, MUST contain the Subject's zip or postal information. | [Section 3.2.3](#323-authentication-of-individual-identity) |
+| `streetAddress`                | NOT RECOMMENDED | If present, MUST contain the Subject's street address information. | [Section 3.2.3](#323-authentication-of-individual-identity) |
+| `organizationName`             | NOT RECOMMENDED | If present, MUST contain the Subject's name or DBA. | [Section 3.2.3](#323-authentication-of-individual-identity) |
+| `surname`                      | MUST            | The Subject's surname. | [Section 3.2.3](#323-authentication-of-individual-identity) |
+| `givenName`                    | MUST            | The Subject's given name. | [Section 3.2.3](#323-authentication-of-individual-identity) |
+| `organizationalUnitName`       | -               | -           | -           |
+| \ \ \ \ _Prior to 2022-09-01_  | NOT RECOMMENDED | If present, the CA MUST implement a process that prevents an OU attribute from including a name, DBA, tradename, trademark, address, location, or other text that refers to a specific natural person or Legal Entity unless the CA has verified this information. | [Section 3.2](#32-initial-identity-validation) |
+| \ \ \ \ _Effective 2022-09-01_ | MUST NOT        | -           | -           |
+| `commonName`                   | NOT RECOMMENDED | If present, MUST contain a single IP address or Fully-Qualified Domain Name that is one of the values contained in the Certificate's `subjectAltName` extension. | |
+| Any other attribute            | NOT RECOMMENDED | -           | See [Section 7.1.4.3](#7143-other-subject-attributes) |
+
+In addition, the following requirements apply to `subject` Attributes:
+
+  * `subject` Attributes MUST NOT contain only metadata such as '.', '-', and ' ' (i.e. space) characters, and/or any other indication that the value is absent, incomplete, or not applicable.
+  * `subject` Attributes other than `commonName` MUST NOT include a Domain Name or IP Address.
```

Comment:
Resuming the discussion here since #39 has been closed...

Can you explain a bit more how you believe this is an existing BR requirement? I don't see any obligation in the BRs to validate Domain Names that are not contained in the SAN/CN according to 3.2.2.4.

Reply:
@CBonnell 

From 7.1.2.4

> CAs **SHALL NOT include a Domain Name or IP Address in a Subject attribute** except as specified in Section 3.2.2.4 or Section 3.2.2.5.

From 3.2.2.4
> Note: FQDNs may be listed in Subscriber Certificates using dNSNames in the subjectAltName extension or in Subordinate CA Certificates via dNSNames in permittedSubtrees within the Name Constraints extension.

From 9.6.1:
> implemented a procedure for verifying that the Applicant either had the right to use, or had control of, the Domain Name(s) and IP address(es) **listed in the Certificate’s subject field and subjectAltName extension** (or, only in the case of Domain Names, was delegated such right or control by someone who had such right to use or control);

Reply:
Isn't the note in 3.2.2.4 clarifying that validation is only required for the SAN dNSNames or permittedSubtrees field of the NC extension? I don't believe it's prohibiting the inclusion of values that conform to Domain Name syntax in other fields. If it did, the current language would prohibit the inclusion of Domain Names in CN.

More fundamentally, any prohibition on Domain Names or IP Addresses from appearing in Subject attributes would have the logical conclusion that any OV/EV Subscriber Certificate today with a countryName value of "US" is mis-issued unless the Applicant asserted control of the entire ".us" TLD. I don't believe that's the state of the WebPKI today, so the current ballot language is more restrictive than the current requirements.

Reply:
> Isn't the note in 3.2.2.4 clarifying that validation is only required for the SAN dNSNames or permittedSubtrees field of the NC extension?

I've already read it as an upper-bound, not a minimum.

>  If it did, the current language would prohibit the inclusion of Domain Names in CN.

Well, we already limit things, by requiring the CN contain the SAN, right?


I think we're in agreement that the language, as written, is internally inconsistent. However, I'm not sure how you can hold the current position, since it seems to require ignoring the clear language of 7.1.2.4 and 9.6.1. I wasn't sure: Were you trying to point out the inconsistency (which is trying to be addressed), or were you trying to argue that the overall interpretation holds consistent with the requirements?

> I don't believe that's the state of the WebPKI today, so the current ballot language is more restrictive than the current requirements.

"More restrictive" is an interesting choice, because it seems the comparison is between "Don't do it" and "validate the authorization if you do it" (unless you ignore 7.1.2.4/9.6.1). That is, it's only more restrictive if we believe that CAs are presently validating that C=US means the subject is entitled to issue for all of the `.us` TLD, and that this would forbid it.

But I think that gets to your above point, that you haven't seen the requirements of 7.1.2.4/9.6.1 as "validate every domain name in a(ny) Subject Field", but rather "validate any domain names in (these) Subject fields". That's not what we have presently written, though, but perhaps you can help me understand how you reach that conclusion?

For example, digging in the older BRs, such as when the CA/B Forum allowed `domainComponent`, which it did until [Ballot 148](https://cabforum.org/2015/04/02/ballot-148-issuer-field-correction/) in [Version 1.2.5 (redline)](https://cabforum.org/wp-content/uploads/BRv1.2.5-redlined.pdf)

Of course, even if you look back that far, the language was unambiguous:
> CAs SHALL NOT include a Domain Name in a Subject attribute except as specified in Sections 9.2.1
and 9.2.2(a) below 

The refactor to the BRs (in v.1.3.0) changed that language to the present form (referencing 3.2.2.4 / 3.2.2.5), but you can see from the discussion, the intent was to preserve the existing requirements.

Reply:
> But I think that gets to your above point, that you haven't seen the requirements of 7.1.2.4/9.6.1 as "validate every domain name in a(ny) Subject Field", but rather "validate any domain names in (these) Subject fields". That's not what we have presently written, though, but perhaps you can help me understand how you reach that conclusion?

Although it sounds like you don't agree, I believe the Note in 3.2.2.4 makes it clear that validation is only required for values contained in the SAN/NC. The language in 7.1.2.4 and 9.6.1 references 3.2.2.4, so, while not entirely clear, the BRs are not stating there is an obligation to validate other Subject attribute values per 3.2.2.4/5, only those that contain values echoed in the SAN/NC. Otherwise, as I pointed out previously, if the current requirements are to be read that *every* Subject attribute value that contains a Domain Name or IP Address must be validated by 3.2.2.4/3.2.2.5, then virtually every OV/EV certificate issued today is mis-issued due to the lack of validation of the countryName value as a Domain Name (or other attributes, for that matter). I imagine the vast majority BR readers would find this to be a quite surprising interpretation.

In drafting the ballot language, I think it would be useful to first have a goal in mind. If I understand the concern correctly, we want to ensure that CN values undergo 3.2.2.4/3.2.2.5 validation, as the CN-ID may be used for hostname identification. I am unaware of any standard that defines other attributes as suitable for hostname identification. In other words, if a client implementation is extracting the hostname from the organizationName attribute, then that implementation is flawed. Therefore, I'm not seeing a reason to mandate 3.2.2.4/3.2.2.5 for any attribute besides CN. Can you clarify what your goal is here?

Reply:
> The language in 7.1.2.4 and 9.6.1 references 3.2.2.4

No, 9.6.1 doesn’t. And I hope I showed how 7.1.2.4 evolved and was unquestionably clear that Domain Names are not permitted in Subjects.

I think we’ve reached the limit of our disagreement though. You seem to be ignoring multiple sections - and historic versions that were even less ambiguous.

The example you highlight is a great example, and absolutely demonstrates the process we hope CAs would apply to all requirements. However, using that example to decide then that the requirements don’t apply - rather than the requirements need clarity - is absolutely the anti-pattern profiles is trying to address. I hope I’ve sufficiently demonstrated how this requirement evolved, how the bug was introduced, but how your interpretation was not and is not supported by the text.

> If I understand the concern correctly, we want to ensure that CN values undergo 3.2.2.4/3.2.2.5 validation, as the CN-ID may be used for hostname identification.

Can you explain how this conclusion was reached? I feel like we’ve had this conversation several times on the calls now, and it’s been clarified this is not the case, so I’m wondering how we’re still reaching this point.

The argument being advancing here is the usual inconsistent argument: namely, we cannot say that only the application validation matters AND that the non-application relevant information is useful to users, while holding the interpretation and path forward being advanced. If only application validation matters, then all the Subject attributes should be forbidden, and we should be using critical SANs. If we say that Subject attributes matter to users, then we need them to be meaningful and not-misleading, and including URIs or Domain Names (as both end up being prohibited) in Subject attributes defeats that.

This is not about protecting misguided applications, but protecting misguided users.

What are the implications of a user attempting to access `example.com`, receiving a certificate for `evil.example`, _viewing_ the certificate (as many CAs insist they should), and seeing an entity incorporated in Kentucky called `example.com`? Under these proposed rules, that would be perfectly fine - would the user be mislead (yes). This is even more important given CAs’ present ability (which, as a concession to comity, this ballot did not tackle in v1) to issue for arbitrary attributes.

Perhaps it’s best to pivot to have you explain: what concerns you, besides the `countryName`, with this proposed requirement? In the case of CAs, you raised the question “Should we have to validate that SSL.com (the organization) owns `ssl.com` (the domain),” to which the answer seems a clear yes, that doesn’t seem unreasonable. What concerns exist here for Subscribers?

Is it that some CAs are not be able to advertise their services in the Subject field. Past discussions of this requirement concluded such advertisements were a clear violation of this, but you seem to be arguing that no, this was fine, because the requirement isn’t what it plainly says.

Reply:
> No, 9.6.1 doesn’t. And I hope I showed how 7.1.2.4 evolved and was unquestionably clear that Domain Names are not permitted in Subjects.

It's not explicit, but what other section in the BRs describes the validation processes for Domain Names besides 3.2.2.4?

> Can you explain how this conclusion was reached? I feel like we’ve had this conversation several times on the calls now, and it’s been clarified this is not the case, so I’m wondering how we’re still reaching this point.

The minutes of https://lists.cabforum.org/pipermail/validation/2021-November/001728.html are clear that validation-sc participants do not necessarily share this view.

> The argument being advancing here is the usual inconsistent argument: namely, we cannot say that only the application validation matters AND that the non-application relevant information is useful to users, while holding the interpretation and path forward being advanced. If only application validation matters, then all the Subject attributes should be forbidden, and we should be using critical SANs. If we say that Subject attributes matter to users, then we need them to be meaningful and not-misleading, and including URIs or Domain Names (as both end up being prohibited) in Subject attributes defeats that.

The BRs define the validation requirements for many subject attribute fields, so I'm not sure how you reach the conclusion that values are "non-meaningful" and "misleading" if they conform to Domain Name syntax but weren't validated per 3.2.2.4. And even for those fields that aren't explicitly enumerated in the BRs, the CA has an obligation to disclose its validation processes in its CP/CPS.

> What are the implications of a user attempting to access example.com, receiving a certificate for evil.example, viewing the certificate (as many CAs insist they should), and seeing an entity incorporated in Kentucky called example.com? Under these proposed rules, that would be perfectly fine - would the user be mislead (yes). This is even more important given CAs’ present ability (which, as a concession to comity, this ballot did not tackle in v1) to issue for arbitrary attributes.

Is there a legal entity registered in KY named "example.com" and they have requested a certificate? If so, then there is nothing misleading; the certificate contains an accurate representation of the validated identity information. If not, then the CA has failed in properly validating the legal entity and the certificate is mis-issued.

> Perhaps it’s best to pivot to have you explain: what concerns you, besides the countryName, with this proposed requirement? In the case of CAs, you raised the question “Should we have to validate that SSL.com (the organization) owns ssl.com (the domain),” to which the answer seems a clear yes, that doesn’t seem unreasonable. What concerns exist here for Subscribers?

Any attribute that contains a value conforming to Domain Name or IP Address syntax now must be validated. `countryName` best conveys the point I am making, but it applies to any other attribute. `localityName` value of "St.Louis" or "Pittsburgh"? Needs to be validated per 3.2.2.4. `postalCode` of 15222? 3.2.2.5 applies, as it's the decimal or hex representation of an IP address. And the list goes on.

> Is it that some CAs are not be able to advertise their services in the Subject field. Past discussions of this requirement concluded such advertisements were a clear violation of this, but you seem to be arguing that no, this was fine, because the requirement isn’t what it plainly says.

I think there's a middle ground that needs to be established. It's not reasonable to include CA branding in OU, but neither is it reasonable to expect every attribute value that so happens to match Domain Name/IP Address syntax to be validated per 3.2.2.4/3.2.2.5. Which is why I think it's useful to find agreement on the goal and then flesh out the requirements from there.

Reply:
> It's not explicit, but what other section in the BRs describes the validation processes for Domain Names besides 3.2.2.4?

Sorry, I'm having trouble following your argument here.

It sounds like the argument being made is that despite 9.6.1's language, namely:
> implemented a procedure for verifying that the Applicant either had the right to use, or had control of, **the Domain Name(s) and IP address(es) listed in the Certificate’s subject field** and subjectAltName extension (or, only in the case of Domain Names, was delegated such right or control by someone who had such right to use or control);

Is a reference to 3.2.2.4, and that 3.2.2.4's statement, namely:
> Note: FQDNs may be listed in Subscriber Certificates using dNSNames in the subjectAltName extension or in Subordinate CA Certificates via dNSNames in permittedSubtrees within the Name Constraints extension.

Should be read and concluded that CAs can issue certificates with Domain Name(s) and IP Address(es), without _any validation according to 3.2.2.4_, because 3.2.2.4 doesn't say they can't?

I want to make sure the argument being made is clear, because that's how it is presently reading. That is, the combined statement that the "may" is being interpreted not as an upper-bound but a lower-bound, and that 9.6.1 isn't a restriction

> > Can you explain how this conclusion was reached? I feel like we’ve had this conversation several times on the calls now, and it’s been clarified this is not the case, so I’m wondering how we’re still reaching this point.

This didn't answer my question.

To be clear, my question was trying to understand how you concluded
> we want to ensure that CN values undergo 3.2.2.4/3.2.2.5 validation, as the CN-ID may be used for hostname identification.

As the goal. As it's not the goal, and that's been repeatedly stated.

> The BRs define the validation requirements for many subject attribute fields, so I'm not sure how you reach the conclusion that values are "non-meaningful" and "misleading" if they conform to Domain Name syntax but weren't validated per 3.2.2.4. And even for those fields that aren't explicitly enumerated in the BRs, the CA has an obligation to disclose its validation processes in its CP/CPS.

I'm not sure that saying "The rules are whatever you want them to be" somehow makes something meaningful and non-misleading. But I suspect this is a bigger distraction, although it does reveal a further interesting area of CA misinterpretation of requirements and goals.

> Is there a legal entity registered in KY named "example.com" and they have requested a certificate? If so, then there is nothing misleading; the certificate contains an accurate representation of the validated identity information. If not, then the CA has failed in properly validating the legal entity and the certificate is mis-issued.

This is the fundamental question. It's interesting to note how many member CAs had a different take, in the case of Stripe Kentucky, thus leading to the revocation of that certificate. There, the argument was that "misleading" is not something to be assessed as "Technically correct" (from the data source), but rather, "did not comport to user expectations and understanding of this information".

Which is my point: CAs are wanting to hold both sides when it's advantageous, and the goal - of the existing language, where this is already a requirement (as demonstrated through its history) - was to provide something clear, consistent, and which applied a restrictive interpretation to limit edge cases.

So let's start with the end goal, which is stated in our existing 9.6.1
> That, at the time of issuance, the CA implemented a procedure for verifying that the Applicant either had the right to use, or had control of, the Domain Name(s) and IP address(es) listed in the Certificate’s subject field and subjectAltName extension (or, only in the case of Domain Names, was delegated such right or control by someone who had such right to use or control);

All Subject fields. From there, we can see if there's the middle-ground to be had. For example, `countryName`, `localityName`, `stateOrProvinceName`, `postalCode` - these are all fields with external registries, so perhaps these are fine to blanket allowlist. The "threat actor" there would be the entity assigning the naming, which would be a sovereign state.

That leaves (in the explicitly permitted fields), `streetAddress`, `givenName`, `surname`, `organizationName`, all of which have a degree of flexibility and control by a third-party, although which MAY have some supervision from a sovereign state ("it varies"). If I changed my legal name to `google.com`, I doubt a reasonable or rationale argument could made that it would not be misleading.

`organizationalUnitName` and other subject attributes, by virtue of the lack of consistent validation across CAs, are wholly out, because users cannot and should not be expected to internalize the constantly changing (within a CA) and entirely inconsistent (within the industry) set of requirements.

This would be a weakening of our current BR requirements, but in a way that hopefully addresses those edge cases you've highlighted. Does this seem workable?

Reply:
@CBonnell It’s unclear: is this a refinement to the allowlist approach I outlined, or an attempted replacement for it?

I think we have ample experience that a “default-allow” approach is a bit unworkable. I would be equally curious to hear from other browsers, such as @clintwilson or @benwilson-mozilla, on whether there’s agreement that this is a current (but buggy) requirement to be fixed, since that helps set our baseline. And that baseline likely needs to start with a “default-deny”, particularly with respect to arbitrary (unspecified) fields.

I hopefully clarified the goal, which is to achieve our longstanding language in a clear and unambiguous fashion, which is prohibiting Domain Names or IP addresses in Subject fields, except as explicitly allowed.

It sounds like your suggestion is to just make sure to specify what format a Domain Name or IP address looks like in the Subject, such as making sure “two or more” domain labels. This would allow “Aerospace Museum, llc”, but disallow “aerospace.museum, llc” as intended.

Reply:
Yeah I interpret this to be a current, but clearly not-fully-intended, requirement that all subject fields be validated. That should get fixed, and an allowlist seems like a good way to do it, e.g. where subject fields have specific validation requirements, those should supersede. 
Refining the definition of Domain Name formatting also seems tractable, but I think I'm leaning towards the default-deny + allowlist clarification angle.

Reply:
@clintwilson So to double check:
- Deny Domains and IPs from the Subject, as a blanket base rule
- Clarify that if you validated the explicitly enumerated fields using the defined method, they’re allowed exceptions to the above
  - For ex, if your corporate/legal name is “example.com”, then potentially example.com would be allowed without validating you’re IANA (the owner/operator of example.com)
  - Or do we require 3.2.2.4 validation for this edge, but allow its inclusion of validation succeeds?
- Clarify the definition of Domain and IP so exactly what shape of things being forbidden is clear/unambiguous

Reply:
I'd be OK with allowing "aerospace.museum" in a subject field if it also passed domain validation of BR 3.2.2.4 but not allow it if it didn't. Is that the question?

Reply:
@BenWilson-Mozilla I believe the question is whether to require (domain) validation at all, or simply allow `aerospace.museum` if that is the company name.

It sounds like there’s some coalescing around the idea that arbitrary Subject attributes (those not covered by the BRs) should forbid domain names and IPs. The question remaining is mostly to do around Applicant-controlled fields (like `givenName`, `surname`, `organizationName`, and in the extreme case, `streetAddress`), and whether to require 3.2.2.4 validation if they are “Domain/IP shaped” (that is, the clarified terminology), or simply allow them without further validation/restriction, as long as that’s how they exist in the business source.

Reply:
> @clintwilson So to double check:
> 
>     * Deny Domains and IPs from the Subject, as a blanket base rule
+1
> 
>     * Clarify that if you validated the explicitly enumerated fields using the defined method, they’re allowed exceptions to the above
+1
>       
>       * For ex, if your corporate/legal name is “example.com”, then potentially example.com would be allowed without validating you’re IANA (the owner/operator of example.com)
>       * Or do we require 3.2.2.4 validation for this edge, but allow its inclusion of validation succeeds?
I'd certainly prefer we continue to require 3.2.2.4/3.2.2.5 validation for this type of edge. I'd be curious to know from CAs what I'm overlooking if that were made clearer, i.e. what unintended consequences/impacts it might have, but based on what I know now it seems a clear better outcome.
> 
>     * Clarify the definition of Domain and IP so exactly what shape of things being forbidden is clear/unambiguous
+1

Reply:
Personally (i.e. input is needed from others), I think that BOTH organization vetting through government business registries AND 3.2.2.4 validation would be the way to go for organization names that are also domain names before they are allowed in other subject attribute fields. I don't see a need for anyone to put IP addresses in other applicant-controlled fields, so that should just be forbidden.

Reply:
OK, this is super helpful, thanks @clintwilson and @BenWilson-Mozilla !

I agree, it’s definitely something worth further discussion during the review. And @CBonnell has rightfully pointed out areas the current language is overbroad, so we can definitely work to address those.

@CBonnell I’ll try to make an update to the language here to clarify, based on the excellent points you’ve raised, and we’ll continue to discuss. But I think the above shows a bit of directional alignment to the goal, and how we fix the existing (but buggy/overly restrictive, as you point out) language.

Reply:
@CBonnell 

This turned out to be rather devilish, so thanks again for raising this. In looking through UTS#46 and related, it understandably is complex to validate this in a reliable, automated fashion ([the best kind](https://www.youtube.com/watch?v=hou0lU8WMgo) of validation).

The options here are either taking an overly broad/restrictive approach, which shunts every 'maybe' into a DV case, which I think is fine in theory, but still somewhat tricky to define as such, or attempt to come up with a nuanced approach that still lets "obvious" badness through. For example, the Numbers Period ( https://unicode-table.com/en/blocks/enclosed-alphanumerics/ - U+2488 - U+249B ) that decompose into fullstops, which are not valid IDNA, but which could still be visually confusable (U+2491 0 . 0 . 1) becomes "10.0.0.1"

I'm still needing to do CT examinations for this data to understand the implications for existing certs, but the current thinking for "potential domain name or IP address" language is

- Two or more sequences of two or more characters separated by FULL STOP (U+002E), FULLWIDTH FULL STOP (U+FF03), IDEOGRAPHIC FULL STOP (U+3002), HALFWIDTH IDEOGRAPHIC FULL STOP (U+FF61)
  - Note: This ends up not being able to prevent the "Numbers Period" character ranges because of their decomposition, unless they're mapped per IDNA
  - Note: This also ends up including/catching "spol. s r.o" (Slovakia), since whitespace is a character (invalid in IDNA, but you don't know until you've processed the string)
  - Note: This ends up (incorrectly) allowing `domainComponent` decomposition, which you highlighted in https://github.com/sleevi/cabforum-docs/pull/36#discussion_r859176437 
- Any sequence of characters that may be successfully parsed according to the textual representation of an IPv6 address, per [RFC 4291, Section 2.2](https://datatracker.ietf.org/doc/html/rfc4291#section-2.2)
- Four sequences of digits separated by FULL STOP (U+002E), FULLWIDTH FULL STOP (U+FF03), IDEOGRAPHIC FULL STOP (U+3002), HALFWIDTH IDEOGRAPHIC FULL STOP (U+FF61)


Alternatively, if we accept that (malicious) confusables are possible, for as long as "attacker controlled" strings are present in the Subject (e.g. before any explicit MUST NOT of any other value), then an alternative is just to repurpose the language from [Ballot SC48v2](https://cabforum.org/2021/07/22/ballot-sc48-v2-domain-name-and-ip-address-encoding/), focusing on just the ASCII minimal encodings, and saying no instances (outside of `organizationName`, per above)

What are your thoughts, while I dig into this?